_posts/2024-04-20-websockets2.md

```diff
@@ -0,0 +1,1861 @@
+
+```bash
+docker run -d --name postgres-server -v postgres_volume:/var/lib/postresql/data -p 5432:5432 -e POSTGRES_PASSWORD=password postgres:alpine
+```
+
+Here we have the `run` command in detached mode by passing the `-d` flag, we also create a volume in the `/var/lib/postresql/data` path, this will help persist Postgres data even if we accidentally delete the container. If you're on Windows or Mac, this path may be different.
+
+We also pass the internal and external ports as `5432` using the `-p` flag and lastly create the password for the `postgres` user by passing `-e POSTGRES_PASSWORD=password`. Finally at the end of the command we pass the image name, in our case `postgres:alpine`.
+
+We can confirm that our container is running by `docker ps`. Next, we'll create the tables we need for our application, we start by accessing the container command running:
+
+```bash
+docker exec -it postgres-server bash                                                                
+```
+We then run `psql -U postgres` to get access to postgres within the container:
+
+```bash 
+root@8d7c752f20e5:/# psql -U postgres
+psql (12.17 (Debian 12.17-1.pgdg110+1))
+Type "help" for help.
+
+postgres=# 
+```
+We can now paste pass the following SQL command to create our database:
+
+```sql
+CREATE DATABASE websocket;
+```
+We can also connect to it by running:
+
+```bash
+postgres=# \c websocket
+You are now connected to database "websocket" as user "postgres".
+websocket=# 
+```
+Once connected let's create the necessary `users` and `rooms` tables:
+
+```sql
+CREATE TABLE users (
+    id UUID PRIMARY KEY,
+    name VARCHAR(255) NOT NULL
+);
+
+CREATE TABLE rooms (
+    id UUID PRIMARY KEY,
+    name VARCHAR(255) NOT NULL
+);
+```
+Both these tables are similar, with an `id` of type `UUID` marked as the `PRIMARY KEY` and a name of type `VARCHAR(255)` marked as `NOT NULL`.
+
+Lastly let's also create the `messages` table:
+
+```sql
+CREATE TABLE messages (
+    id UUID PRIMARY KEY,
+    message TEXT NOT NULL,
+    time TIMESTAMP DEFAULT NOW(),
+    user_id UUID REFERENCES users (id),
+    room_id UUID REFERENCES rooms (id)
+);
+```
+It also has an `id` of type `UUID`, `message` of type `TEXT`, `time` of type `TIMESTAMP`, and it references the `users` and `rooms` `id`s as `user_id` and `room_id` respectively.
+
+### 4.2. Redis
+We'll also be using the official Redis image from docker hub, we can fetch it by running the following command:
+```bash
+docker pull redis
+```
+Setting up the Redis container is much simpler compared to Postgres, this can be done with the following single command:
+```bash
+docker run -d --name redis-server -p 6379:6379 redis    
+```
+Here we use the docker run command again, however, we pass the `name` as `redis-server` and port as `6379`.
+
+## 5. Skunk
+In this section, we'll implement the protocols necessary for interacting with Postgres in our application using `Skunk`.
```

Comment:
Let's add a link to the article I've just published:
https://blog.rockthejvm.com/skunk-complete-guide/